cfq-iosched: do not force idling for sync workload

From: Konstantin Khlebnikov <khlebnikov@openvz.org>

revert v2.6.32-108-gc04645e
blkio: Wait on sync-noidle queue even if rq_noidle = 1
by Vivek Goyal <vgoyal@redhat.com>

and piece of v2.6.32-rc5-486-g8e55063
cfq-iosched: fix corner cases in idling logic
by Corrado Zoccolo <czoccolo@gmail.com>


fix perfomance degradation for massive write-fsync pattern:
# sysbench --test=fileio --file-num=1 --file-total-size=1G --file-fsync-all=on \
--file-test-mode=seqwr --max-time=10 --file-block-size=4096 --max-requests=0 run

http://bugzilla.openvz.org/show_bug.cgi?id=1622

Signed-off-by: Konstantin Khlebnikov <khlebnikov@openvz.org>
---

 block/cfq-iosched.c |    9 +--------
 1 files changed, 1 insertions(+), 8 deletions(-)

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 023f4e6..b68b633 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -3333,14 +3333,7 @@ static void cfq_completed_request(struct request_queue *q, struct request *rq)
 		else if (sync && cfqq_empty &&
 			 !cfq_close_cooperator(cfqd, cfqq)) {
 			cfqd->noidle_tree_requires_idle |= !rq_noidle(rq);
-			/*
-			 * Idling is enabled for SYNC_WORKLOAD.
-			 * SYNC_NOIDLE_WORKLOAD idles at the end of the tree
-			 * only if we processed at least one !rq_noidle request
-			 */
-			if (cfqd->serving_type == SYNC_WORKLOAD
-			    || cfqd->noidle_tree_requires_idle
-			    || cfqq->cfqg->nr_cfqq == 1)
+			if (cfqd->noidle_tree_requires_idle)
 				cfq_arm_slice_timer(cfqd);
 		}
 	}
